File: 2021/day2/day2.exs
File.read("input.txt")
|> elem(1)
|> String.trim()
|> String.split("\n")
|> Enum.map(fn line ->
  [dir, len] = String.split(line)
  {len, _} = Integer.parse(len)
  {dir, len}
end)
|> Enum.reduce([0, 0, 0], fn {dir, len}, [x, y, a] ->
  case dir do
    "down" -> [x, y, a + len]
    "up" -> [x, y, a - len]
    "forward" -> [x + len, y + a * len, a]
  end
end)
|> Enum.take(2)
|> Enum.product()
|> IO.puts()

File: 2021/day2/day2.py
with open("input.txt", "r") as f:
    input = f.read()

input = input.strip().splitlines()

x, y, a = 0, 0, 0
for line in input:
    dir, len = line.split()
    len = int(len)

    match dir:
        case "down":
            a += len
        case "up":
            a -= len
        case "forward":
            x += len
            y += a * len

print(x * y)
